sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -620,6 +628,106 @@ def _next_id(self):
     return str(self._last_id)
 
 
+class CachingMaterializingStateHandler(object):
+  """ A State handler which retrieves and caches state. """
+
+  def __init__(self, global_state_cache, underlying_state):
+    self._underlying = underlying_state
+    self._state_cache = global_state_cache
+    self._context = threading.local()
+
+  @contextlib.contextmanager
+  def process_instruction_id(self, bundle_id, cache_tokens):
+    if getattr(self._context, 'cache_token', None) is not None:
+      raise RuntimeError(
+          'Cache tokens already set to %s' % self._context.cache_token)
+    # TODO Also handle cache tokens for side input, if present
+    user_state_cache_token = None
+    for cache_token_struct in cache_tokens:
+      if cache_token_struct.HasField("user_state"):
+        # There should only be one user state token present
+        assert not user_state_cache_token
+        user_state_cache_token = cache_token_struct.token
+    try:
+      self._context.cache_token = user_state_cache_token
+      with self._underlying.process_instruction_id(bundle_id):
+        yield
+    finally:
+      self._context.cache_token = None
+
+  def blocking_get(self, state_key, coder, is_cached=False):
+    if not self._should_be_cached(is_cached):
+      # no cache / tokens, can't do a lookup/store in the cache
+      return self._materialize_iter(state_key, coder)
+    # Cache lookup
+    cache_state_key = self._convert_to_cache_key(state_key)
+    cached_value = self._state_cache.get(cache_state_key,
+                                         self._context.cache_token)
+    if cached_value is None:
+      # Cache miss, need to retrieve from the Runner
+      materialized = cached_value = []
```

Comment:
This makes it impossible to have state larger than fits into memory, whereas before one could page it in. Also, this means the whole thing must be read before the first element gets processed which could be a performance regression.

Reply:
Possibly one idea would be to only cache the first page (with the reasoning that for something so large one may be OK with repeated fetches).

Reply:
For the cache to work effectively, we want to cache the whole value. Note that if the cache is disabled, we still retain the old behavior of only materializing a page at a time.

Yes, it has the drawback of the value to fit into memory. The design currently doesn't do memory estimation but keeps a fixed upper limit of items. If we move to a memory estimation, then also reflecting this here would make sense.

Reply:
The pagination is intended for when the value is too big to send, which probably means it's too big to cache (especially many of them given the current number-of-items bound). I don't think one would want to turn off all caching to avoid caching just these.

Reply:
I think caching only the first page makes sense. From looking at the Runner code, it looks like continuation tokens are not used currently. It seems like it is fair to defer this until they are used; also considering that the state caching is disabled by default.